<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/images/SN_web_lightmode.png" width="300" alt="cognitiveclass.ai logo">
</center>

<h1 align=center><font size = 5>Assignment: Notebook for Graded Assessment</font></h1>


# Introduction

Using this Python notebook you will:

1.  Understand three Chicago datasets
2.  Load the three datasets into three tables in a SQLIte database
3.  Execute SQL queries to answer assignment questions


## Understand the datasets

To complete the assignment problems in this notebook you will be using three datasets that are available on the city of Chicago's Data Portal:

1.  <a href="https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Socioeconomic Indicators in Chicago</a>
2.  <a href="https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Chicago Public Schools</a>
3.  <a href="https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Chicago Crime Data</a>

### 1. Socioeconomic Indicators in Chicago

This dataset contains a selection of six socioeconomic indicators of public health significance and a “hardship index,” for each Chicago community area, for the years 2008 – 2012.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2](https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

### 2. Chicago Public Schools

This dataset shows all school level performance data used to create CPS School Report Cards for the 2011-2012 school year. This dataset is provided by the city of Chicago's Data Portal.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t](https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

### 3. Chicago Crime Data

This dataset reflects reported incidents of crime (with the exception of murders where data exists for each victim) that occurred in the City of Chicago from 2001 to present, minus the most recent seven days.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2](https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)


### Download the datasets

This assignment requires you to have these three tables populated with a subset of the whole datasets.

In many cases the dataset to be analyzed is available as a .CSV (comma separated values) file, perhaps on the internet. 

Use the links below to read the data files using the Pandas library. 

* Chicago Census Data

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

* Chicago Public Schools

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

* Chicago Crime Data

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

**NOTE:** Ensure you use the datasets available on the links above instead of directly from the Chicago Data Portal. The versions linked here are subsets of the original datasets and have some of the column names modified to be more database friendly which will make it easier to complete this assignment.


### Store the datasets in database tables

To analyze the data using SQL, it first needs to be loaded into SQLite DB.
We will create three tables in as under:

1.  **CENSUS_DATA**
2.  **CHICAGO_PUBLIC_SCHOOLS**
3.  **CHICAGO_CRIME_DATA**


Load the `pandas` and `sqlite3` libraries and establish a connection to `FinalDB.db`


In [1]:
import pandas as pd
import csv, sqlite3

con = sqlite3.connect(database='FinalDB.db')
cur = con.cursor()

Load the SQL magic module


In [2]:
%load_ext sql

Use `Pandas` to load the data available in the links above to dataframes. Use these dataframes to load data on to the database `FinalDB.db` as required tables.


In [3]:
df = pd.read_csv("ChicagoCensusData.csv")
df.to_sql(name="CENSUS_DATA", con=con, if_exists='replace', index=False, method="multi")

df = pd.read_csv("ChicagoCrimeData.csv")
df.to_sql(name="CHICAGO_CRIME_DATA", con=con, if_exists='replace', index=False, method="multi")

df = pd.read_csv("ChicagoPublicSchools.csv")
df.to_sql(name="CHICAGO_PUBLIC_SCHOOLS_DATA", con=con, if_exists='replace', index=False)

566

Establish a connection between SQL magic module and the database `FinalDB.db`


In [4]:
%sql sqlite:///FinalDB.db

You can now proceed to the the following questions. Please note that a graded assignment will follow this lab and there will be a question on each of the problems stated below. It can be from the answer you received or the code you write for this problem. Therefore, please keep a note of both your codes as well as the response you generate.


## Problems

Now write and execute SQL queries to solve assignment problems

### Problem 1

##### Find the total number of crimes recorded in the CRIME table.


In [25]:
%sql select distinct Primary_Type, Description from CHICAGO_CRIME_DATA \
group by Primary_Type

 * sqlite:///FinalDB.db
Done.


[('ARSON', 'AGGRAVATED'),
 ('ASSAULT', 'SIMPLE'),
 ('BATTERY', 'SIMPLE'),
 ('BURGLARY', 'UNLAWFUL ENTRY'),
 ('CONCEALED CARRY LICENSE VIOLATION', 'OTHER'),
 ('CRIM SEXUAL ASSAULT', 'ATTEMPT NON-AGGRAVATED'),
 ('CRIMINAL DAMAGE', 'TO VEHICLE'),
 ('CRIMINAL TRESPASS', 'TO STATE SUP LAND'),
 ('DECEPTIVE PRACTICE', 'CREDIT CARD FRAUD'),
 ('DOMESTIC VIOLENCE', 'DOMESTIC VIOLENCE'),
 ('GAMBLING', 'GAME/DICE'),
 ('HOMICIDE', 'FIRST DEGREE MURDER'),
 ('HUMAN TRAFFICKING', 'INVOLUNTARY SERVITUDE'),
 ('INTERFERENCE WITH PUBLIC OFFICER', 'RESIST/OBSTRUCT/DISARM OFFICER'),
 ('INTIMIDATION', 'INTIMIDATION'),
 ('KIDNAPPING', 'CHILD ABDUCTION/STRANGER'),
 ('LIQUOR LAW VIOLATION', 'SELL/GIVE/DEL LIQUOR TO MINOR'),
 ('MOTOR VEHICLE THEFT', 'AUTOMOBILE'),
 ('NARCOTICS', 'POSS: CANNABIS 30GMS OR LESS'),
 ('NON - CRIMINAL', 'FOID - REVOCATION'),
 ('NON-CRIMINAL', 'LOST PASSPORT'),
 ('NON-CRIMINAL (SUBJECT SPECIFIED)', 'NOTIFICATION OF CIVIL NO CONTACT ORDER'),
 ('OBSCENITY', 'OBSCENE MATTER'),
 ('OFFENSE INVOLVING CHILDREN', 'AGG SEX ASSLT OF CHILD FAM MBR'),
 ('OTHER NARCOTIC VIOLATION', 'INTOXICATING COMPOUNDS'),
 ('OTHER OFFENSE', 'TELEPHONE THREAT'),
 ('PROSTITUTION', 'CALL OPERATION'),
 ('PUBLIC INDECENCY', 'LICENSED PREMISE'),
 ('PUBLIC PEACE VIOLATION', 'RECKLESS CONDUCT'),
 ('RITUALISM', 'AGG RIT MUT: HANDS/FIST/FEET SERIOUS INJURY'),
 ('ROBBERY', 'ARMED: HANDGUN'),
 ('SEX OFFENSE', 'AGG CRIMINAL SEXUAL ABUSE'),
 ('STALKING', 'CYBERSTALKING'),
 ('THEFT', 'FROM BUILDING'),
 ('WEAPONS VIOLATION', 'UNLAWFUL POSS OF HANDGUN')]

In [9]:
%sql SELECT name, type, length(type) FROM PRAGMA_TABLE_INFO('CHICAGO_CRIME_DATA')

 * sqlite:///FinalDB.db
Done.


[('ID', 'INTEGER', 7),
 ('CASE_NUMBER', 'TEXT', 4),
 ('DATE', 'TEXT', 4),
 ('BLOCK', 'TEXT', 4),
 ('IUCR', 'TEXT', 4),
 ('PRIMARY_TYPE', 'TEXT', 4),
 ('DESCRIPTION', 'TEXT', 4),
 ('LOCATION_DESCRIPTION', 'TEXT', 4),
 ('ARREST', 'INTEGER', 7),
 ('DOMESTIC', 'INTEGER', 7),
 ('BEAT', 'INTEGER', 7),
 ('DISTRICT', 'INTEGER', 7),
 ('WARD', 'REAL', 4),
 ('COMMUNITY_AREA_NUMBER', 'REAL', 4),
 ('FBICODE', 'TEXT', 4),
 ('X_COORDINATE', 'REAL', 4),
 ('Y_COORDINATE', 'REAL', 4),
 ('YEAR', 'INTEGER', 7),
 ('LATITUDE', 'REAL', 4),
 ('LONGITUDE', 'REAL', 4),
 ('LOCATION', 'TEXT', 4)]

In [10]:
%sql SELECT count(Case_Number) FROM CHICAGO_CRIME_DATA

 * sqlite:///FinalDB.db
Done.


[(533,)]

### Problem 2

##### List community area names and numbers with per capita income less than 11000.


In [11]:
%sql SELECT name, type, length(type) FROM PRAGMA_TABLE_INFO('CENSUS_DATA')

 * sqlite:///FinalDB.db
Done.


[('COMMUNITY_AREA_NUMBER', 'REAL', 4),
 ('COMMUNITY_AREA_NAME', 'TEXT', 4),
 ('PERCENT_OF_HOUSING_CROWDED', 'REAL', 4),
 ('PERCENT_HOUSEHOLDS_BELOW_POVERTY', 'REAL', 4),
 ('PERCENT_AGED_16__UNEMPLOYED', 'REAL', 4),
 ('PERCENT_AGED_25__WITHOUT_HIGH_SCHOOL_DIPLOMA', 'REAL', 4),
 ('PERCENT_AGED_UNDER_18_OR_OVER_64', 'REAL', 4),
 ('PER_CAPITA_INCOME', 'INTEGER', 7),
 ('HARDSHIP_INDEX', 'REAL', 4)]

In [12]:
%sql select Community_Area_Name, Per_Capita_Income FROM CENSUS_DATA where \
Per_Capita_Income < 11000

 * sqlite:///FinalDB.db
Done.


[('West Garfield Park', 10934),
 ('South Lawndale', 10402),
 ('Fuller Park', 10432),
 ('Riverdale', 8201)]

### Problem 3

##### List all case numbers for crimes involving minors?(children are not considered minors for the purposes of crime analysis) 


In [26]:
%%sql
select
  case_number, 
  primary_type,
  description
from
  CHICAGO_CRIME_DATA
where
  description like '%minor%'

 * sqlite:///FinalDB.db
Done.


[('HL266884', 'LIQUOR LAW VIOLATION', 'SELL/GIVE/DEL LIQUOR TO MINOR'),
 ('HK238408', 'LIQUOR LAW VIOLATION', 'ILLEGAL CONSUMPTION BY MINOR')]

### Problem 4

##### List all kidnapping crimes involving a child?


In [28]:
%sql select case_number, Primary_Type, Description from CHICAGO_CRIME_DATA \
where Primary_Type = 'KIDNAPPING'

 * sqlite:///FinalDB.db
Done.


[('HN144152', 'KIDNAPPING', 'CHILD ABDUCTION/STRANGER')]

### Problem 5

##### List the kind of crimes that were recorded at schools. (No repetitions)


In [42]:
%sql select Distinct Primary_Type, Location_Description FROM CHICAGO_CRIME_DATA \
where Location_Description like '%School%' group by Primary_Type

 * sqlite:///FinalDB.db
Done.


[('ASSAULT', 'SCHOOL, PUBLIC, GROUNDS'),
 ('BATTERY', 'SCHOOL, PUBLIC, GROUNDS'),
 ('CRIMINAL DAMAGE', 'SCHOOL, PUBLIC, GROUNDS'),
 ('CRIMINAL TRESPASS', 'SCHOOL, PUBLIC, GROUNDS'),
 ('NARCOTICS', 'SCHOOL, PUBLIC, GROUNDS'),
 ('PUBLIC PEACE VIOLATION', 'SCHOOL, PRIVATE, BUILDING')]

### Problem 6

##### List the type of schools along with the average safety score for each type.


In [43]:
%sql select * from CHICAGO_PUBLIC_SCHOOLS_DATA limit 3

 * sqlite:///FinalDB.db
Done.


[(610038, 'Abraham Lincoln Elementary School', 'ES', '615 W Kemper Pl', 'Chicago', 'IL', 60614, '(773) 534-5720', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610038.pdf', 'Fullerton Elementary Network', 'NORTH-NORTHWEST SIDE COLLABORATIVE', 'No', 'Standard', 'Not on Probation', 'Level 1', 'Yes', 'Very Strong', 99.0, 'Very Strong', '99', 'Strong', 74.0, 'Strong', 66.0, 'Weak', '65', 'Strong', '70', 'Strong', '56', 'Average', '47', '96.00%', 2.0, '96.40%', '95.80%', '80.1', '43.3', '89.6', '84.9', '60.7', '62.6', '81.9', '85.2', '52', '62.4', '66.3', '77.9', 69.7, 64.4, 0.2, 0.9, 'Yellow', 'Green', '67.1', '54.5', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 813, 33, 'NDA', 1171699.458, 1915829.428, 41.92449696, -87.64452163, 7, 'LINCOLN PARK', 43, 18, '(41.92449696, -87.64452163)'),
 (610281, 'Adam Clayton Powell Paideia Community Academy Elementary School', 'ES', '7511 S South Shore Dr', 'Chicago', 'IL', 60649, '(773) 535-6650', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610281.pdf', 'Skyway Elementary Network', 'SOUTH SIDE COLLABORATIVE', 'No', 'Track_E', 'Not on Probation', 'Level 1', 'No', 'Average', 54.0, 'Strong', '66', 'Strong', 74.0, 'Very Strong', 84.0, 'Weak', '63', 'Strong', '76', 'Weak', '46', 'Average', '50', '95.60%', 15.7, '95.30%', '100.00%', '62.4', '51.7', '21.9', '15.1', '29', '42.8', '38.5', '27.4', '44.8', '42.7', '14.1', '34.4', 16.8, 16.5, 0.7, 1.4, 'Green', 'Green', '17.2', '27.3', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 521, 46, 'NDA', 1196129.985, 1856209.466, 41.76032435, -87.55673627, 43, 'SOUTH SHORE', 7, 4, '(41.76032435, -87.55673627)'),
 (610185, 'Adlai E Stevenson Elementary School', 'ES', '8010 S Kostner Ave', 'Chicago', 'IL', 60652, '(773) 535-2280', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610185.pdf', 'Midway Elementary Network', 'SOUTHWEST SIDE COLLABORATIVE', 'No', 'Standard', 'Not on Probation', 'Level 2', 'No', 'Strong', 61.0, 'NDA', 'NDA', 'Average', 50.0, 'Weak', 36.0, 'Weak', 'NDA', 'NDA', 'NDA', 'Average', '47', 'Weak', '41', '95.70%', 2.3, '94.70%', '98.30%', '53.7', '26.6', '38.3', '34.7', '43.7', '57.3', '48.8', '39.2', '46.8', '44', '7.5', '21.9', 18.3, 15.5, -0.9, -1.0, 'Red', 'Red', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 1324, 44, 'NDA', 1148427.165, 1851012.215, 41.74711093, -87.73170248, 70, 'ASHBURN', 13, 8, '(41.74711093, -87.73170248)')]

In [55]:
%sql select distinct 'Elementary, Middle, or High School', AVG(Safety_Score) from CHICAGO_PUBLIC_SCHOOLS_DATA \
group by 'Elementary, Middle, or High School'

 * sqlite:///FinalDB.db
Done.


[('Elementary, Middle, or High School', 49.50487329434698)]

### Problem 7

##### List 5 community areas with highest % of households below poverty line


In [56]:
%sql select community_area_name, percent_households_below_poverty from CENSUS_DATA \
order by percent_households_below_poverty desc limit 5

 * sqlite:///FinalDB.db
Done.


[('Riverdale', 56.5),
 ('Fuller Park', 51.2),
 ('Englewood', 46.6),
 ('North Lawndale', 43.1),
 ('East Garfield Park', 42.4)]

### Problem 8

##### Which community area is most crime prone? Display the coumminty area number only.


In [58]:
%sql select community_area_number, count(community_area_number) as freq from CHICAGO_CRIME_DATA \
group by community_area_number order by count(community_area_number) desc limit 1

 * sqlite:///FinalDB.db
Done.


[(25.0, 43)]


Double-click **here** for a hint

<!--
Query for the 'community area number' that has most number of incidents
-->


### Problem 9

##### Use a sub-query to find the name of the community area with highest hardship index


In [59]:
%sql select community_area_name, hardship_index from CENSUS_DATA \
where hardship_index = (select MAX(hardship_index) from CENSUS_DATA)

 * sqlite:///FinalDB.db
Done.


[('Riverdale', 98.0)]

### Problem 10

##### Use a sub-query to determine the Community Area Name with most number of crimes?


In [61]:
%sql select community_area_name, community_area_number from CENSUS_DATA \
where community_area_number = (select community_area_number from CHICAGO_CRIME_DATA \
                               group by community_area_number \
                                order by count(community_area_number) desc limit 1)

 * sqlite:///FinalDB.db
Done.


[('Austin', 25.0)]

## Author(s)

<h4> Hima Vasudevan </h4>
<h4> Rav Ahuja </h4>
<h4> Ramesh Sannreddy </h4>

## Contribtuor(s)

<h4> Malika Singla </h4>
<h4>Abhishek Gagneja</h4>

## Change log

| Date       | Version | Changed by        | Change Description                             |
| ---------- | ------- | ----------------- | ---------------------------------------------- |
|2023-10-18  | 2.6     | Abhishek Gagneja  | Modified instruction set |
| 2022-03-04 | 2.5     | Lakshmi Holla     | Changed markdown.                   |
| 2021-05-19 | 2.4     | Lakshmi Holla     | Updated the question                           |
| 2021-04-30 | 2.3     | Malika Singla     | Updated the libraries                          |
| 2021-01-15 | 2.2     | Rav Ahuja         | Removed problem 11 and fixed changelog         |
| 2020-11-25 | 2.1     | Ramesh Sannareddy | Updated the problem statements, and datasets   |
| 2020-09-05 | 2.0     | Malika Singla     | Moved lab to course repo in GitLab             |
| 2018-07-18 | 1.0     | Rav Ahuja         | Several updates including loading instructions |
| 2018-05-04 | 0.1     | Hima Vasudevan    | Created initial version                        |

## <h3 align="center"> © IBM Corporation 2023. All rights reserved. <h3/>
